File: dal-python/tests/test_curvedata.py
"""Tests for curve data factories."""

import dal
import pytest


def _today():
    return dal.Date_(2025, 6, 20)


def _spot():
    return _today().AddDays(2)


# ---- DiscountPWLF_New ----

def test_discount_pwlf_new():
    """DiscountPWLF_New constructs a flat-forward discount curve."""
    knot_dates = [_spot(), _spot().AddDays(1825)]  # spot + 5 years
    fwd_rates = [0.04, 0.04]
    curve = dal.DiscountPWLF_New("flat", "USD", knot_dates, fwd_rates)
    assert curve is not None  # nosec B101 - pytest assertions are intentional


def test_discount_pwlf_new_multiple_knots():
    """DiscountPWLF_New works with multiple knot points."""
    knot_dates = []
    fwd_rates = []
    for y in [0, 1, 2, 5, 10]:
        knot_dates.append(_spot().AddDays(y * 365))
        fwd_rates.append(0.05)
    curve = dal.DiscountPWLF_New("multi", "USD", knot_dates, fwd_rates)
    assert curve is not None  # nosec B101 - pytest assertions are intentional


def test_discount_pwlf_new_with_base():
    """DiscountPWLF_New accepts an optional base discount curve."""
    knot_dates = [_spot(), _spot().AddDays(365)]
    fwd_rates = [0.03, 0.03]
    base = dal.DiscountPWLF_New("base", "USD", knot_dates, fwd_rates)
    assert base is not None  # nosec B101 - pytest assertions are intentional

    knot_dates2 = [_spot(), _spot().AddDays(1825)]
    fwd_rates2 = [0.04, 0.04]
    curve = dal.DiscountPWLF_New("bootstrapped", "USD", knot_dates2, fwd_rates2, base=base)
    assert curve is not None  # nosec B101 - pytest assertions are intentional


# ---- DiscountZeroRate_New ----

def test_discount_zero_rate_new_defaults():
    """DiscountZeroRate_New supplies ACT/365F and log-linear defaults."""
    curve = dal.DiscountZeroRate_New(
        "zero_default",
        "USD",
        _today(),
        [_today().AddDays(365), _today().AddDays(730)],
        [0.02, 0.025],
    )

    assert curve is not None  # nosec B101 - pytest assertions are intentional


@pytest.mark.parametrize(
    "scheme",
    [
        dal.LogDfScheme.LOG_LINEAR,
        dal.LogDfScheme.LOG_CUBIC_NATURAL,
        dal.LogDfScheme.MIXED,
    ],
)
def test_discount_zero_rate_new_explicit_options(scheme):
    """DiscountZeroRate_New accepts an explicit day count and every log-DF scheme."""
    curve = dal.DiscountZeroRate_New(
        "zero_explicit",
        "USD",
        _today(),
        [_today().AddDays(360), _today().AddDays(720), _today().AddDays(1080)],
        [0.02, 0.025, 0.03],
        day_count=dal.DayBasis_New("ACT_360"),
        log_df_scheme=scheme,
    )

    assert curve is not None  # nosec B101 - pytest assertions are intentional


def test_discount_zero_rate_new_with_base():
    """DiscountZeroRate_New accepts a base curve for multiplicative layering."""
    node_dates = [_today().AddDays(365), _today().AddDays(730)]
    base = dal.DiscountZeroRate_New("base", "USD", _today(), node_dates, [0.01, 0.01])
    curve = dal.DiscountZeroRate_New(
        "zero_spread",
        "USD",
        _today(),
        node_dates,
        [0.02, 0.02],
        base=base,
    )

    assert curve is not None  # nosec B101 - pytest assertions are intentional


@pytest.mark.parametrize(
    "node_dates,zero_rates",
    [
        (lambda today: [today], [0.02]),
        (lambda today: [today.AddDays(365)], []),
        (lambda today: [today.AddDays(365)], [float("nan")]),
    ],
)
def test_discount_zero_rate_new_rejects_invalid_inputs(node_dates, zero_rates):
    """Native ZERO_RATE validation is translated to Python exceptions."""
    today = _today()

    with pytest.raises(RuntimeError, match="zero-rate discount curve"):
        dal.DiscountZeroRate_New(
            "invalid_zero",
            "USD",
            today,
            node_dates(today),
            zero_rates,
        )


# ---- CurveBlock_New (simple) ----

def test_curve_block_new_simple():
    """CurveBlock_New constructs a CurveBlock_ from a single discount curve."""
    knot_dates = [_spot(), _spot().AddDays(1825)]
    fwd_rates = [0.04, 0.04]
    dc = dal.DiscountPWLF_New("ois", "USD", knot_dates, fwd_rates)

    block = dal.CurveBlock_New(dc)
    assert block is not None  # nosec B101 - pytest assertions are intentional


def test_curve_block_new_simple_with_basis():
    """CurveBlock_New accepts an optional Libor day basis."""
    knot_dates = [_spot(), _spot().AddDays(1825)]
    fwd_rates = [0.04, 0.04]
    dc = dal.DiscountPWLF_New("ois", "USD", knot_dates, fwd_rates)

    block = dal.CurveBlock_New(dc, libor_basis=dal.DayBasis_New("ACT_360"))
    assert block is not None  # nosec B101 - pytest assertions are intentional


# ---- CurveBlock_New (full) ----

def test_curve_block_new_full():
    """CurveBlock_New constructs a CurveBlock_ from discount and forward curve maps."""
    knot_dates = [_spot(), _spot().AddDays(1825)]

    ois_curve = dal.DiscountPWLF_New("ois", "USD", knot_dates, [0.04, 0.04])
    libor_curve = dal.DiscountPWLF_New("libor", "USD", knot_dates, [0.045, 0.045])

    discounts = {dal.CollateralType_OIS(): ois_curve}
    forwards = {dal.PeriodLength_New("3M"): libor_curve}

    block = dal.CurveBlock_New("usd", "USD", discounts, forwards, dal.DayBasis_New("ACT_365F"))
    assert block is not None  # nosec B101 - pytest assertions are intentional

File: dal-public/tests/test_curvedata.cpp
//
// Created by wegamekinglc on 2026/6/20.
//

#include <gtest/gtest.h>

#include <dal-public/src/curvedata.hpp>
#include <dal-public/src/curveprotocol.hpp>
#include <dal/curve/ycconst.hpp>
#include <dal/curve/yclogdf.hpp>
#include <dal/curve/ycpwlf.hpp>

using Dal::CollateralType_OIS;
using Dal::CurveBlockNew;
using Dal::Date_;
using Dal::DayBasis_New;
using Dal::DiscountLogDFNew;
using Dal::DiscountPWCNew;
using Dal::DiscountPWLFNew;
using Dal::DiscountZeroRate_;
using Dal::DiscountZeroRateNew;
using Dal::Handle_;
using Dal::LogDfScheme_;
using Dal::PeriodLength_New;
using Dal::String_;
using Dal::Vector_;

namespace {

    Date_ Today() { return Date_(2025, 6, 20); }
    Date_ Spot() { return Today().AddDays(2); }

} // namespace

// DiscountPWLFNew

TEST(CurveDataTest, TestDiscountPWLFNew) {
    Vector_<Date_> knotDates;
    knotDates.push_back(Spot());
    knotDates.push_back(Spot().AddDays(1825)); // 5 years
    Vector_<> fwdRates;
    fwdRates.push_back(0.04);
    fwdRates.push_back(0.04);
    auto curve = DiscountPWLFNew(String_("flat"), String_("USD"), knotDates, fwdRates);
    ASSERT_TRUE(curve != nullptr);
}

TEST(CurveDataTest, TestDiscountPWLFNewMultipleKnots) {
    Vector_<Date_> knotDates;
    Vector_<> fwdRates;
    for (int y : {0, 1, 2, 5, 10}) {
        knotDates.push_back(Spot().AddDays(y * 365));
        fwdRates.push_back(0.05);
    }
    auto curve = DiscountPWLFNew(String_("multi"), String_("USD"), knotDates, fwdRates);
    ASSERT_TRUE(curve != nullptr);
}

TEST(CurveDataTest, TestDiscountPWLFNewWithBase) {
    Vector_<Date_> knotDates;
    knotDates.push_back(Spot());
    knotDates.push_back(Spot().AddDays(365));
    Vector_<> fwdRates;
    fwdRates.push_back(0.03);
    fwdRates.push_back(0.03);
    auto base = DiscountPWLFNew(String_("base"), String_("USD"), knotDates, fwdRates);
    ASSERT_TRUE(base != nullptr);

    Vector_<Date_> knotDates2;
    knotDates2.push_back(Spot());
    knotDates2.push_back(Spot().AddDays(1825));
    Vector_<> fwdRates2;
    fwdRates2.push_back(0.04);
    fwdRates2.push_back(0.04);
    auto curve = DiscountPWLFNew(String_("bootstrapped"), String_("USD"), knotDates2, fwdRates2, base);
    ASSERT_TRUE(curve != nullptr);
}

TEST(CurveDataTest, TestCompleteCurveFactoriesExposeExactReconstructionState) {
    const Vector_<Date_> knotDates{
        Today().AddDays(30),
        Today().AddDays(365),
        Today().AddDays(730),
        Today().AddDays(1095),
    };
    const Vector_<> right{0.01, 0.02, 0.03, 0.04};
    const Vector_<> left{0.011, 0.021, 0.031, 0.041};
    const auto base = DiscountPWCNew("base", "USD", knotDates, right);

    const auto pwlf = DiscountPWLFNew("pwlf", "USD", knotDates, left, right, base);
    const auto* typedPwlf = dynamic_cast<const Dal::Tape::DiscountPWLF_<double>*>(pwlf.get());
    ASSERT_NE(typedPwlf, nullptr);
    ASSERT_EQ(typedPwlf->KnotDates(), knotDates);
    ASSERT_EQ(typedPwlf->FLeft(), left);
    ASSERT_EQ(typedPwlf->FRight(), right);
    ASSERT_EQ(typedPwlf->Base().get(), base.get());

    const Vector_<Date_> logDates = Dal::Vector::Join(Vector_<Date_>{Today()}, knotDates);
    const Vector_<> logDf{0.0, -0.001, -0.02, -0.05, -0.09};
    Dal::MappedDiscountCurveOptions_ options;
    options.dayCount_ = DayBasis_New("ACT_360");
    options.logDfScheme_ = LogDfScheme_::Value_::MIXED;
    options.base_ = base;
    const auto mapped = DiscountLogDFNew("log", "USD", logDates, logDf, options);
    const auto* typedMapped = dynamic_cast<const Dal::DiscountLogDF_*>(mapped.get());
    ASSERT_NE(typedMapped, nullptr);
    ASSERT_EQ(typedMapped->NodeDates(), logDates);
    ASSERT_EQ(typedMapped->NodeLogDF(), logDf);
    ASSERT_EQ(typedMapped->DayCount().String(), String_("ACT_360"));
    ASSERT_EQ(typedMapped->Scheme(), LogDfScheme_::Value_::MIXED);
    ASSERT_EQ(typedMapped->Base().get(), base.get());
}

// DiscountZeroRateNew

TEST(CurveDataTest, TestDiscountZeroRateNewDefaults) {
    const Vector_<Date_> nodeDates{Today().AddDays(365), Today().AddDays(730), Today().AddDays(1095)};
    const Vector_<> zeroRates{0.02, 0.022, 0.024};

    const auto curve = DiscountZeroRateNew("zero", "USD", Today(), nodeDates, zeroRates);
    const auto* typed = dynamic_cast<const DiscountZeroRate_*>(curve.get());

    ASSERT_NE(typed, nullptr);
    ASSERT_EQ(typed->AnchorDate(), Today());
    ASSERT_EQ(typed->NodeDates(), nodeDates);
    ASSERT_EQ(typed->NodeZeroRates(), zeroRates);
    ASSERT_EQ(typed->DayCount().String(), String_("ACT_365F"));
    ASSERT_EQ(typed->Scheme(), LogDfScheme_::Value_::LOG_LINEAR);
}

TEST(CurveDataTest, TestDiscountZeroRateNewOptionsAndBaseMatchCoreFactory) {
    const Vector_<Date_> nodeDates{Today().AddDays(365), Today().AddDays(730), Today().AddDays(1095)};
    const Vector_<> baseRates{0.01, 0.011, 0.012};
    const Vector_<> spreadRates{0.004, 0.005, 0.006};
    const auto base = DiscountZeroRateNew("base", "USD", Today(), nodeDates, baseRates);
    const Date_ query = Today().AddDays(600);

    for (const auto scheme : {LogDfScheme_::Value_::LOG_LINEAR, LogDfScheme_::Value_::LOG_CUBIC_NATURAL, LogDfScheme_::Value_::MIXED}) {
        const auto curve = DiscountZeroRateNew("spread", "USD", Today(), nodeDates, spreadRates, DayBasis_New("ACT_360"), scheme, base);
        const Handle_<Dal::DiscountCurve_> core(
            Dal::NewDiscountZeroRate("core", "USD", Today(), nodeDates, spreadRates, DayBasis_New("ACT_360"), scheme, base));
        const auto* typed = dynamic_cast<const DiscountZeroRate_*>(curve.get());

        ASSERT_NE(typed, nullptr);
        ASSERT_EQ(typed->DayCount().String(), String_("ACT_360"));
        ASSERT_EQ(typed->Scheme(), scheme);
        ASSERT_NEAR((*curve)(Today(), query), (*core)(Today(), query), 1e-14);
    }
}

// CurveBlockNew (simple overload)

TEST(CurveDataTest, TestCurveBlockNewSimple) {
    Vector_<Date_> knotDates;
    knotDates.push_back(Spot());
    knotDates.push_back(Spot().AddDays(1825));
    Vector_<> fwdRates;
    fwdRates.push_back(0.04);
    fwdRates.push_back(0.04);
    auto dc = DiscountPWLFNew(String_("ois"), String_("USD"), knotDates, fwdRates);

    auto block = CurveBlockNew(dc);
    ASSERT_TRUE(block != nullptr);
}

TEST(CurveDataTest, TestCurveBlockNewSimpleWithBasis) {
    Vector_<Date_> knotDates;
    knotDates.push_back(Spot());
    knotDates.push_back(Spot().AddDays(1825));
    Vector_<> fwdRates;
    fwdRates.push_back(0.04);
    fwdRates.push_back(0.04);
    auto dc = DiscountPWLFNew(String_("ois"), String_("USD"), knotDates, fwdRates);

    auto block = CurveBlockNew(dc, DayBasis_New("ACT_360"));
    ASSERT_TRUE(block != nullptr);
}

// CurveBlockNew (full overload)

TEST(CurveDataTest, TestCurveBlockNewFull) {
    // Build two discount curves: OIS and Libor 3M
    Vector_<Date_> knotDates;
    knotDates.push_back(Spot());
    knotDates.push_back(Spot().AddDays(1825));
    Vector_<> oisRates;
    oisRates.push_back(0.04);
    oisRates.push_back(0.04);
    auto oisCurve = DiscountPWLFNew(String_("ois"), String_("USD"), knotDates, oisRates);

    Vector_<> liborRates;
    liborRates.push_back(0.045);
    liborRates.push_back(0.045);
    auto liborCurve = DiscountPWLFNew(String_("libor"), String_("USD"), knotDates, liborRates);

    std::map<Dal::CollateralType_, Handle_<Dal::DiscountCurve_>> discounts;
    discounts[CollateralType_OIS()] = oisCurve;

    std::map<Dal::PeriodLength_, Handle_<Dal::DiscountCurve_>> forwards;
    forwards[PeriodLength_New("3M")] = liborCurve;

    auto block = CurveBlockNew(String_("usd"), String_("USD"), discounts, forwards, DayBasis_New("ACT_365F"));
    ASSERT_TRUE(block != nullptr);
    ASSERT_EQ(block->DiscountCurves().size(), 1);
    ASSERT_EQ(block->ForwardCurves().size(), 1);
    ASSERT_EQ(block->LiborBasis().String(), String_("ACT_365F"));
}
